### PR TITLE
Make axolotl Flatpak installer work again

### PR DESCRIPTION
**Pull request by olof-nord**
Update the local flatpak to include the new rust library, and bundle Electron at buildtime.

To build:

`make build-dependencies-flatpak-web`
`make install-flatpak-web`

`flatpak run org.nanuc.Axolotl`

The flatpak dependencies are also updated to the latest version.

Build a single binary is built with `astilectron-bundler` which consists of the following:

1) axolotl-web js/css/html bundle
2) Electron
3) compiled axolotl go binary

The `crayfish` and `zkgroup` dependencies are built and included with the flatpak install.

The Electron startup code is slightly adjusted, as to start with the bundled Electron if configured so.
For indicating this, a new environment variable is introduced: `AXOLOTL_ELECTRON_BUNDLED`

TODO:

* [x] Look into how to start axolotl the best way
* [x] Look into how to use the Flatpak base Electron instead of downloading electron with astielectron during first startup
* [x] Use Flatpak bundled Electron build provided at build-time and not downloaded at first startup
* [x] Decide how to handle the Electron startup flags (use Chromium auto-detection: use Wayland if possible)
* [x] Look into Electron missing window controls

### CODE REVIEW

**Comment by olof-nord** [comment]
Logs for current flatpak build.

This line seems to indicate some issues with Electron:

`time="2022/01/29 19:32:17" level=debug msg="[axolotl-electron] main: creating window failed: context canceled"`

```log
➜  axolotl git:(flatpak-with-makefile-integration-take-two) flatpak run org.nanuc.Axolotl 
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] setup config"
time="2022-01-29T19:32:16+01:00" level=info msg="[axolotl] Starting Signal for Ubuntu version 1.0.9"
time="2022-01-29T19:32:16+01:00" level=info msg="[axolotl] Setup completed"
time="2022-01-29T19:32:16+01:00" level=info msg="[axolotl-electron] creating astilelectron with the following switches: [--disable-dev-shm-usage --no-sandbox]"
time="2022-01-29T19:32:16+01:00" level=info msg="astikit: starting worker..."
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] Run Backend"
time="2022-01-29T19:32:16+01:00" level=info msg="[axolotl] Axolotl server started"
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] openDb: /home/olof/.local/share/textsecure.nanuc/db"
time="2022-01-29T19:32:16+01:00" level=info msg="[axolotl] Starting axolotl ws"
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] NewDataStore"
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] Using axolotl-web path ./axolotl-web/dist"
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] NewDataStore finished"
time="2022-01-29T19:32:16+01:00" level=info msg="[axolotl] Loading Chats"
time="2022-01-29T19:32:16+01:00" level=info msg="[axolotl] Db setup finished"
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] Start Session after Decryption"
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] starting Signal connection"
time="2022-01-29T19:32:16+01:00" level=debug msg="[textsecure] loading config"
time="2022-01-29T19:32:16+01:00" level=info msg="[textsecure-crayfish] Starting crayfish-backend"
time="2022-01-29T19:32:16+01:00" level=debug msg="[axolotl] config path:  /home/olof/.config/textsecure.nanuc"
time="2022/01/29 19:32:16" level=debug msg="[texsecure] register Device"
time="2022/01/29 19:32:16" level=debug msg="[textsecure] loading config"
time="2022/01/29 19:32:16" level=debug msg="[axolotl] config path:  /home/olof/.config/textsecure.nanuc"
time="2022/01/29 19:32:16" level=debug msg="[textsecure] Crayfish registration starting"
time="2022/01/29 19:32:16" level=debug msg="[axolotl] config path:  /home/olof/.config/textsecure.nanuc"
time="2022/01/29 19:32:16" level=debug msg="[axolotl] Opening Dialog: getPhoneNumber"
time="2022/01/29 19:32:16" level=debug msg="[axolotl-ws] send request getPhoneNumber"
time="2022/01/29 19:32:16" level=debug msg="[textsecure-crayfish-ws] BackendStartListening"
time="2022/01/29 19:32:16" level=debug msg="[textsecure-crayfish-ws] websocket connecting to crayfish-server"
time="2022/01/29 19:32:16" level=error msg="dial tcp [::1]:9081: connect: connection refused"
time="2022/01/29 19:32:16" level=error msg="[textsecure-crayfish-ws] Failed to start listening" error="dial tcp [::1]:9081: connect: connection refused"
time="2022/01/29 19:32:16" level=info msg="[crayfish] starting crayfish 7bbaf1665234ce8dd3455ded2164b107fae7f972\n"
time="2022/01/29 19:32:16" level=info msg="[crayfish] Starting libsignal-service web socket at ws://127.0.0.1:9081/libsignal\n"
time="2022-01-29T19:32:17+01:00" level=error msg="'/home/olof/.config/textsecure.nanuc/electron/vendor/electron-linux-amd64/electron' exited with code: -1"
time="2022-01-29T19:32:17+01:00" level=info msg="astikit: stopping worker..."
time="2022/01/29 19:32:17" level=debug msg="[axolotl-electron] main: creating window failed: context canceled"
time="2022/01/29 19:32:17" level=debug msg="[axolotl-electron] open dev tools false"
time="2022-01-29T19:32:17+01:00" level=info msg="astikit: worker is now waiting..."
time="2022-01-29T19:32:17+01:00" level=error msg="accept tcp 127.0.0.1:39273: use of closed network connection while TCP accepting"
time="2022/01/29 19:32:17" level=info msg="[axolotl] ending axolotl"
➜  axolotl git:(flatpak-with-makefile-integration-take-two) 

```

**Comment by nuehm-arno** [comment]
Hey @olof-nord ,

do you think, it's possible to implement these changes in a custom built of Electron for Pinephone distros?

https://gitlab.gnome.org/World/Phosh/squeekboard/-/issues/244

**Comment by olof-nord** [comment]
The challenges which using Electron inside Flatpak with the astilectron-bundler are ironed out in a separate [mini-project](https://gitlab.com/olof-nord/go-astilectron-flatpak-demo) as to allow a faster iteration and fewer moving parts.

The initial issue with Flatpak/Wayland is solved and I have just asked @asticode for a [second opinion](https://github.com/asticode/go-astilectron-bundler/issues/100) about some details in regards to using astilectron-bundler with Flatpak.

**Comment by olof-nord** [comment]
The `astilectron-bundler` Flatpak provisioning errors are ironed out, but Electron + Wayland comes without window controls.
Window controls are the X to close the window for example.

This is a feature which seems to be included with Electron with this PR, https://github.com/electron/electron/pull/29618 and is potentially included with the recent Electron 17 release. However when trying it out locally the application crashes for me.

As release 17 of the Electron project is so new (just a Feb 1st, -22), it is also not completely clear to me if this feature is actually included or not: there is not much documentation about this yet.

**Comment by nuehm-arno** [comment]
Hey @olof-nord ,

impressive work here. One question regarding [main.go](https://github.com/nanu-c/axolotl/pull/733/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261):

You removed the if-condition whether Electron should start with the Ozone platform activated or not (`electronSwitches`). Why is that? Is every OS, on which Axolotl is running, using Wayland now? Or is there another switch somewhere in the Electron universe which I missed?

**Comment by olof-nord** [comment]
Thanks for the reminder about the Electron flags!

I added them all unconditionally when I was testing out other things, now I brought the Wayland check back.

I would argue that we always add the Wayland flags if we detect a Wayland session though, disregarding of which ARCH we are on.

**Comment by vchernin** [comment]
> However when trying it out locally the application crashes for me

Random drive by commenter here, how exactly did you test it? What was the crash message?

FYI you should only need `--ozone-platform-hint=auto`, not any of the other `ozone` flags in recent Electron/Chromium versions. It enables native Wayland if Chromium/Electron thinks it's available, you shouldn't need to detect it here.

**Comment by olof-nord** [comment]
> > However when trying it out locally the application crashes for me
> 
> Random drive by commenter here, how exactly did you test it? What was the crash message?
> 
> FYI you should only need `--ozone-platform-hint=auto`, not any of the other `ozone` flags in recent Electron/Chromium versions. It enables native Wayland if Chromium/Electron thinks it's available, you shouldn't need to detect it here.

As much as I want this to work, at least with Electron 16.0.8, 17.0.0 and 18.0.0-alpha.1, it does not.

Here an excerpt from the startup logs, using the [demo](https://gitlab.com/olof-nord/go-astilectron-flatpak-demo) project based on the go-astilectron-demo project. I find no Electron documentation regarding this flag, and sparse information online about the "auto" mode in general.

Perhaps there is some further configuration needed alongside this? Perhaps Flatpak doesnt support something?
Perhaps some additional Flatpak finish-args are needed?

I have not much to work with here, other than the "context canceled" message which seems to be the standard error message when the app crashes.

From what I can tell, this would be a convenience method, which in this case actually brings problems rather than convenience: as the XDG_SESSION_TYPE check does the trick, I see keeping that in place as the only working solution at the moment.

```
2022/02/05 20:07:58 Electron has already been provisioned to version 18.0.0-alpha.1, moving on...
2022/02/05 20:07:58 Listening...
2022/02/05 20:07:58 Executing...
2022/02/05 20:07:58 Starting cmd /home/olof/.var/app/com.gitlab.go-astilectron-flatpak-demo/data/vendor/electron-linux-amd64/electron /home/olof/.var/app/com.gitlab.go-astilectron-flatpak-demo/data/vendor/astilectron/main.js 127.0.0.1:37009 true --no-sandbox --ozone-platform-hint=auto
2022/02/05 20:07:58 astikit: received signal child exited
2022/02/05 20:07:58 '/home/olof/.var/app/com.gitlab.go-astilectron-flatpak-demo/data/vendor/electron-linux-amd64/electron' exited with code: -1
2022/02/05 20:07:58 App has crashed
2022/02/05 20:07:58 Stopping...
2022/02/05 20:07:58 astikit: stopping worker...
2022/02/05 20:07:58 Closing...
2022/02/05 20:07:58 accept tcp 127.0.0.1:37009: use of closed network connection while TCP accepting
2022/02/05 20:07:58 Stopping...
2022/02/05 20:07:58 running bootstrap failed: creating window failed: context canceled
```

**Comment by olof-nord** [comment]
I am quite proud to say that with the review from @nuehm-arno and the input from @vchernin, this work is ready to merge.

@nanu-c, could you validate that the Snap build still works as before? Anything else open/to discuss before merging?

File: snap/snapcraft.yaml
Comment (nanu-c) [inline snap/snapcraft.yaml:83]: That's a cool solution.

**Review by nanu-c — APPROVED** [review]

**Comment by nanu-c** [comment]
The snap still works, i was able to build and run one. If you want to try, it's here.
[axolotl_1.0.9_amd64.zip](https://github.com/nanu-c/axolotl/files/8009972/axolotl_1.0.9_amd64.zip)

**Comment by nanu-c** [comment]
So great work Olof, well done! Maybe i try to fix the electron tray icon, or isn't that necessary with flatpak?

**Comment by olof-nord** [comment]
Thanks for the feedback! Yes, the tray icon is not yet in place, that would be next on the list for sure.

I had a look at https://github.com/nanu-c/axolotl/pull/459 last night. I want to dig into it a bit more as to both avoid regressions, but also understand the core problem before suggesting a solution.

**Pull request merged**
